BrainPortal/lib/s3_connection.rb

```diff
@@ -19,52 +18,244 @@
+    @path_name = path_start
+    @endpoint = endpoint
   end
-
+  
+  # Method to translate Mime types to File types
+  # Redo more generally
+  def translate_content_type_to_ftype(ct)
+    ct == "application/x-directory" ? :directory : :regular
+  end
+      
+  # Method to ensure that the paths sent start in the right place
+  def clean_starting_folder_path(path)
+    path_end = path
+    
+    if not path_end.to_s.starts_with? @path_name
+      return Pathname.new(File.join(@path_name, path_end))
+    else
+      return Pathname.new(path_end)
+    end
+  end
+  
   def execute_on_s3 #:nodoc:
      yield self
   end
 
-  # Create a bucket on the current connection.
-  def create_bucket(bucket_name)
+  # Create a bucket on the current connection. Not used, included for completeness
+  def self.create_bucket(bucket_name)
     @bucket.create(bucket_name)
   end
 
   # Returns true if the connection is alive
+  # Need to figure out how to return the actual error message so that one can debug
   def connected?
-    @base.connected?
+    begin
+      testVar = @resource.client.list_objects(bucket: @bucket_name, delimiter: "/")
+      return true 
+    rescue 
+      return false
+    end
+  end
+
+  # Lists all objects by name only directly undder a given path
+  # Returns a hash with files, folders and full path
+  # If path is nil, then it uses only the start_path
+  def list_objects_short(path=nil)
+    path = @path_name + "/" if path.nil?
+    
+    pathClean = clean_starting_folder_path(path).to_s
+    if object_exists?(pathClean)
+      contKey = ""
+      resp = @resource.client.list_objects_v2(bucket: @bucket_name, 
+                                              prefix: pathClean, 
+                                              delimiter: "/")
+                                               
+      filenames = resp.contents.map { |x| {:name => x.key.split("/")[-1],
+                                           :time => x.last_modified,
+                                           :size => x.size,
+                                           :content_type => 'none'} if x.size > 0}.compact
+      filenames.each do |x|
+        new_key = File.join(pathClean,x[:name])
+        respH = @resource.client.head_object({bucket: @bucket_name, key: new_key})
+        x[:content_type] = respH.content_type
+      end
+      folderNames = resp.common_prefixes.map { |x| {:name => x.prefix.split("/")[-1]} }      
+      return {:path => pathClean, :files => filenames, :folders => folderNames}                                   
+     else
+      return {:path => pathClean, :files => [], :folders => []} 
+    end
+  end
+  
+  # List every single object recursively under a given path
+  # returns a list of each object
+  def list_objects_long(path=nil)
+    path = @path_name if path.nil?
+    
+    pathClean = clean_starting_folder_path(path).to_s
+    list_of_objects = Array.new()
+    
+    if object_exists?(pathClean)
+      cont_token = nil
+      while true
```

Comment:
Don't use `while true`, I'm afraid of infinite loops

Reply:
You should learn to embrace you infinite loop while true overlords.